libs/wingc/src/closure_transform.rs

```diff
@@ -0,0 +1,306 @@
+		if self.curr_phase == Phase::Inflight {
+			return expr;
+		}
+
+		// If we encounter a preflight closure, we don't need to
+		// transform it into a preflight class, but we still want to recurse
+		// in case its body contains any inflight closures.
+		if let ExprKind::FunctionClosure(ref func_def) = expr.kind {
+			if func_def.signature.phase != Phase::Inflight {
+				return fold::fold_expr(self, expr);
+			}
+		}
+
+		match expr.kind {
+			ExprKind::FunctionClosure(func_def) => {
+				self.inflight_counter += 1;
+
+				let new_class_name = Symbol {
+					name: format!("$Inflight{}", self.inflight_counter),
+					span: expr.span.clone(),
+				};
+				let handle_name = Symbol {
+					name: "handle".to_string(),
+					span: expr.span.clone(),
+				};
+
+				let class_type_annotation = TypeAnnotation {
+					kind: TypeAnnotationKind::UserDefined(UserDefinedType {
+						root: new_class_name.clone(),
+						fields: vec![],
+					}),
+					span: expr.span.clone(),
+				};
+
+				let class_fields: Vec<ClassField> = vec![];
+				let class_init_params: Vec<FunctionParameter> = vec![];
+
+				let new_func_def = if self.inside_scope_with_this {
+					// If we are inside a class, we transform inflight closures with an extra
+					// `let __parent_this = this;` statement before the class definition, and replace all references
+					// of `this` with `__parent_this` so that they can access the parent class's fields.
+					let mut this_transform = RenameIdentifierTransformer {
+						from: "this",
+						to: PARENT_THIS_NAME,
+					};
+					this_transform.fold_function_definition(func_def)
+				} else {
+					func_def
+				};
+
+				// class_def :=
+				// ```
+				// class <new_class_name> {
+				//   init(<class_init_params>) {<class_init_body>}
+				//   inflight handle() {
+				//     <transformed_def>
+				//   }
+				// }
+				// ```
+				let class_def = Stmt {
+					kind: StmtKind::Class(Class {
+						name: new_class_name.clone(),
+						is_resource: true,
+						initializer: Initializer {
+							signature: FunctionSignature {
+								parameters: class_init_params,
+								return_type: Some(Box::new(class_type_annotation.clone())),
+								phase: Phase::Preflight,
+							},
+							statements: Scope::new(vec![], expr.span.clone()),
+							span: expr.span.clone(),
+						},
+						fields: class_fields,
+						implements: vec![],
+						parent: None,
+						methods: vec![(handle_name.clone(), new_func_def)],
+						inflight_initializer: None,
+					}),
+					idx: self.nearest_stmt_idx,
+					span: expr.span.clone(),
+				};
+
+				// new_class_instance :=
+				// ```
+				// new <new_class_name>();
+				// ```
+				let new_class_instance = Expr::new(
+					ExprKind::New {
+						class: class_type_annotation,
+						arg_list: ArgList {
+							named_args: IndexMap::new(),
+							pos_args: vec![],
+						},
+						obj_id: None,
+						obj_scope: None,
+					},
+					expr.span.clone(),
+				);
+
+				self.class_statements.push(class_def);
```

Comment:
is `class_statements` ever more than a single statement? If not it might be clearer to get rid of this vec.

Reply:
It's possible for it to contain more than one statement if we are calling `fold_statement` on a statement with more than one inflight expression, for example `foo(inflight () => {}, inflight () => {})`

Reply:
:exploding_head: Any chance you can show the "wing" output of that fold so I better understand what's going on here?

Reply:
Sure, so some code like this:
```js
let b = foo(inflight (key: str): str => {
  counter.inc();
}, inflight (key: str): str => {
  counter.dec();
});
```

will get transformed into this:

```js
class $Inflight1 {
  inflight handle(key: str): str {
    counter.inc();
  }
}
class $Inflight2 {
  inflight handle(key: str): str {
    counter.dec();
  }
}
let b = foo(new $Inflight1(), new $Inflight2());
```

I'm having trouble adding this as a unit test though